up-transport-vsomeip/src/transport_inner/transport_inner_engine.rs

```diff
@@ -0,0 +1,852 @@
+/********************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use crate::determine_message_type::RegistrationType;
+use crate::message_conversions::UMessageToVsomeipMessage;
+use crate::storage::UPTransportVsomeipStorage;
+use crate::transport_inner::{
+    UP_CLIENT_VSOMEIP_FN_TAG_APP_EVENT_LOOP, UP_CLIENT_VSOMEIP_FN_TAG_INITIALIZE_NEW_APP_INTERNAL,
+    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL, UP_CLIENT_VSOMEIP_FN_TAG_START_APP,
+    UP_CLIENT_VSOMEIP_TAG,
+};
+use crate::utils::{split_u32_to_u16, split_u32_to_u8};
+use crate::{ApplicationName, ClientId, UeId};
+use cxx::{let_cxx_string, UniquePtr};
+use log::{error, info, trace};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use tokio::runtime::Builder;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::oneshot;
+use up_rust::{UCode, UMessage, UMessageType, UStatus, UUri};
+use vsomeip_sys::glue::{
+    make_application_wrapper, make_payload_wrapper, make_runtime_wrapper, ApplicationWrapper,
+    MessageHandlerFnPtr, RuntimeWrapper,
+};
+use vsomeip_sys::vsomeip;
+
+pub enum TransportCommand {
+    // Primary purpose of a UTransport
+    RegisterListener(
+        UUri,
+        Option<UUri>,
+        RegistrationType,
+        MessageHandlerFnPtr,
+        ApplicationName,
+        Arc<dyn UPTransportVsomeipStorage>,
```

Comment:
it seems strange to me that you need to pass around references to the internal state. IMHO, as the name suggests, _internal_ state should be internal to the component that owns it ...

Reply:
Yeah... I get where you're coming from. I'm not in love with it either. Have you got some ideas?

What lead me down this direction was:
1. We need to allow the extern "C" fns we pass into vsomeip to get access to some of the details of the transport to able to convert vsomeip messages to uProtocol UMessages.
2. We pass the storage into the engine so that we can do book-keeping on the extern "C" fns we use to monitor the availability of the vsomeip application to signal it was sucessfully started (or not)

Reply:
I mitigated 2. a little bit by pass in a trait object of just what's needed.

I could mitigate 1. a little bit more, but I'll leave this up for a future improvement.

Reply:
Calling this mitigated for now